### ISSUE TITLE
runs-on/cache vs actions/cache with magic seem to go to different buckets.

### ISSUE THREAD

**Issue by labeleddata** [runs-on/cache#27]
# tl;dr
Using the runs-on/cache gets much better download times than the magic cache with actions/cache. 
Also, running magic cache multiple times gets a cache hit. Following that with the runs-on/cache is a miss. 

Maybe they are going different places? 

# Details 
We tried the new magic cache, adn tried to debug why the downloads were a bit slow vs runs-on/cache

Speed with magic
<img width="639" alt="image" src="https://github.com/user-attachments/assets/3aa124b7-a325-4ff4-b954-a2874ad25ede" />
Speed with runs on
<img width="600" alt="image" src="https://github.com/user-attachments/assets/cc2341de-ea75-4db3-91a5-5894a97ff608" />

In both situations we have 
```yaml 
jobs:
  python_deps_test:
    runs-on:
      # Runs the job on AWS runners, with a magic s3-cache
      - runs-on=${{ github.run_id }}
      - family=m6in+c6in+r6in
      - cpu=32+128

      - extras=s3-cache

    steps:


      - name: Checkout
        uses: actions/checkout@v3

      # Setup Python environment using the remote reusable composite action
      - name: Setup Python Environment
        uses: ./.github/actions/better-python
        with:
          PIP_CONF: ${{ secrets.PIP_CONF_AZURE }}
          PYTHON_VERSION: ${{ inputs.PYTHON_VERSION }}
``` 

where the caching is happening inside the custom action .

Inside the action we have (using magic cache) 
```yaml
runs:
  using: "composite"
  steps:
    # ----------------------------------------
    # Check if we need to skip Python setup
    # ----------------------------------------
    - uses: runs-on/action@v1
      name: Run this to activate RunsOn magic cache (Google it)


        

    # ----------------------------------------
    # Restore cached venv if available
    # ----------------------------------------
    - name: Restore cached venv
      
      uses: actions/cache@v4
      id: cache-venv
      with:
        path: ${{ inputs.venv_path }}
        key: ${{ runner.os }}-venv-${{ hashFiles(env.PYTHON_REQUIREMENTS_PATH) }}
        restore-keys: |
          ${{ runner.os }}-venv-
```
And using the runs-on cache 
The meat of the action, using runs-on cache is 
```yaml
runs:
  using: "composite"
  steps:
    # ----------------------------------------
    # Check if we need to skip Python setup
    # ----------------------------------------
    - uses: runs-on/action@v1
      name: Run this to activate RunsOn magic cache (Google it)


        

    # ----------------------------------------
    # Restore cached venv if available
    # ----------------------------------------
    - name: Restore cached venv
      
      uses: runs-on/cache@v4
      id: cache-venv
      with:
        path: ${{ inputs.venv_path }}
        key: ${{ runner.os }}-venv-${{ hashFiles(env.PYTHON_REQUIREMENTS_PATH) }}
        restore-keys: |
          ${{ runner.os }}-venv-
```

However, running these one after the other, the magic cache gives us a cache hit  with key `Linux-venv-acfb00df7caf8b5650137b8f855e203b58ab4784bf6518061fc092eb576068e4`
```
Run # If PYTHON_VERSION is not provided, read from .python-version
Run actions/cache@v4
  with:
    path: /tmp/.venv
    key: Linux-venv-acfb00df7caf8b5650137b8f855e203b58ab4784bf6518061fc092eb576068e4
    restore-keys: Linux-venv-
  
    enableCrossOsArchive: false
    fail-on-cache-miss: false
    lookup-only: false
    save-always: false
  env:
    RUNS_ON_AGENT_ARCH: x[64](https://github.com/PrimerAI/nlx-service/actions/runs/12447833866/job/34751747799#step:4:67)
    RUNS_ON_RUNNER_NAME: runs-on--i-0dc2f2[65](https://github.com/PrimerAI/nlx-service/actions/runs/12447833866/job/34751747799#step:4:68)4d01728d1--xrFsOyBsFz
    RUNS_ON_AGENT_USER: runner
    RUNS_ON_S3_BUCKET_CACHE: runs-on-s3bucketcache-xv3gaqedifwa
    RUNS_ON_AWS_REGION: us-west-2
    RUNS_ON_AMI_NAME: runs-on-v2.2-ubuntu22-full-x64-20241216080634
    RUNS_ON_INSTANCE_TYPE: c6in.8xlarge
    RUNS_ON_INSTANCE_ID: i-0dc2f2654d01728d1
    RUNS_ON_INSTANCE_LIFECYCLE: spot
    ACTIONS_RUNNER_HOOK_JOB_STARTED: /opt/runs-on/pre.sh
    ACTIONS_RUNNER_HOOK_JOB_COMPLETED: /opt/runs-on/post.sh
    PYTHON_VERSION: 3.10.14
    PIP_CONF: ***
    extra-index-url :  ***pkgs.dev.azure.com/primer-ai/_packaging/Common/pypi/simple/
    PYTHON_REQUIREMENTS_PATH: requirements/dev.txt
Received 0 of 3141410325 (0.0%), 0.0 MBs/sec
Received 713031[68](https://github.com/PrimerAI/nlx-service/actions/runs/12447833866/job/34751747799#step:4:71) of 3141410325 (2.3%), 32.9 MBs/sec
Received 163577856 of 3141410325 (5.2%), 50.9 MBs/sec
```
while using the runs-on/cache (for the first time) gives us a not found on the same key 

`Cache not found for input keys: Linux-venv-acfb00df7caf8b5650137b8f855e203b58ab4784bf6518061fc092eb576068e4, Linux-venv-` 

```
Run runs-on/cache@v4
  with:
    path: /tmp/.venv
    key: Linux-venv-acfb00df7caf8b5[65](https://github.com/PrimerAI/nlx-service/actions/runs/12447895293/job/34751877408#step:4:68)0137b8f855e203b58ab4784bf6518061fc092eb576068e4
    restore-keys: Linux-venv-
  
    enableCrossOsArchive: false
    fail-on-cache-miss: false
    lookup-only: false
    save-always: false
  env:
    RUNS_ON_AGENT_ARCH: x64
    RUNS_ON_RUNNER_NAME: runs-on--i-0b3501e07cccb2a[66](https://github.com/PrimerAI/nlx-service/actions/runs/12447895293/job/34751877408#step:4:69)--drBcndRRuh
    RUNS_ON_AGENT_USER: runner
    RUNS_ON_S3_BUCKET_CACHE: runs-on-s3bucketcache-xv3gaqedifwa
    RUNS_ON_AWS_REGION: us-west-2
    RUNS_ON_AMI_NAME: runs-on-v2.2-ubuntu22-full-x64-20241216080634
    RUNS_ON_INSTANCE_TYPE: c6in.8xlarge
    RUNS_ON_INSTANCE_ID: i-0b3501e07cccb2a66
    RUNS_ON_INSTANCE_LIFECYCLE: spot
    ACTIONS_RUNNER_HOOK_JOB_STARTED: /opt/runs-on/pre.sh
    ACTIONS_RUNNER_HOOK_JOB_COMPLETED: /opt/runs-on/post.sh
    PYTHON_VERSION: 3.10.14
    PIP_CONF: ***
    extra-index-url :  ***pkgs.dev.azure.com/primer-ai/_packaging/Common/pypi/simple/
    PYTHON_REQUIREMENTS_PATH: requirements/dev.txt
The cache action detected a local S3 bucket cache. Using it.
Cache not found for input keys: Linux-venv-acfb00df7caf8b5650137b8f855e203b58ab4784bf6518061fc092eb5760[68](https://github.com/PrimerAI/nlx-service/actions/runs/12447895293/job/34751877408#step:4:71)e4, Linux-venv-
Run actions/setup-python@v4
```

Running with the runs-on/cache a second time gives a hit, and also downloads at 300MBPs instead of 80

```
n runs-on/cache@v4
  with:
    path: /tmp/.venv
    key: Linux-venv-acfb00df7caf8b5[65](https://github.com/PrimerAI/nlx-service/actions/runs/12447960651/job/34752019243#step:4:68)0137b8f855e203b58ab4784bf6518061fc092eb576068e4
    restore-keys: Linux-venv-
  
    enableCrossOsArchive: false
    fail-on-cache-miss: false
    lookup-only: false
    save-always: false
  env:
    RUNS_ON_AGENT_ARCH: x64
    RUNS_ON_RUNNER_NAME: runs-on--i-04a4f7bef84ff0db5--ZpQiRNRpQM
    RUNS_ON_AGENT_USER: runner
    RUNS_ON_S3_BUCKET_CACHE: runs-on-s3bucketcache-xv3gaqedifwa
    RUNS_ON_AWS_REGION: us-west-2
    RUNS_ON_AMI_NAME: runs-on-v2.2-ubuntu22-full-x64-20241216080634
    RUNS_ON_INSTANCE_TYPE: c6in.8xlarge
    RUNS_ON_INSTANCE_ID: i-04a4f7bef84ff0db5
    RUNS_ON_INSTANCE_LIFECYCLE: spot
    ACTIONS_RUNNER_HOOK_JOB_STARTED: /opt/runs-on/pre.sh
    ACTIONS_RUNNER_HOOK_JOB_COMPLETED: /opt/runs-on/post.sh
    PYTHON_VERSION: 3.10.14
    PIP_CONF: ***
    extra-index-url :  ***pkgs.dev.azure.com/primer-ai/_packaging/Common/pypi/simple/
    PYTHON_REQUIREMENTS_PATH: requirements/dev.txt
  
The cache action detected a local S3 bucket cache. Using it.
Received 3187[67](https://github.com/PrimerAI/nlx-service/actions/runs/12447960651/job/34752019243#step:4:70)104 of 2741244113 (11.6%), 302.8 MBs/sec
Received 754974720 of 2741244113 (27.5%), 359.3 MBs/sec
Received 1191182336 of 2741244113 (43.5%), 377.9 MBs/sec
Received 1627389952 of 2741244113 (59.4%), 387.2 MBs/sec
Received 20635975[68](https://github.com/PrimerAI/nlx-service/actions/runs/12447960651/job/34752019243#step:4:71) of 2741244113 (75.3%), 392.9 MBs/sec
Received 2550136832 of 2741244113 (93.0%), 404.6 MBs/sec
Cache Size: ~2614 MB (2[74](https://github.com/PrimerAI/nlx-service/actions/runs/12447960651/job/34752019243#step:4:77)1244113 B)
```

**Comment by crohr** [comment]
Hi @labeleddata, and thanks for the extensive report!

> Using the runs-on/cache gets much better download times than the magic cache with actions/cache. Also, running magic cache multiple times gets a cache hit. Following that with the runs-on/cache is a miss.
> 
> Maybe they are going different places?

They are not stored at the same path prefix in the bucket yes. In my mind there is no expectation that they are compatible (you use one, or the other), but I'll make that clear on the website.


> Running with the runs-on/cache a second time gives a hit, and also downloads at 300MBPs instead of 80

The underlying mechanism is different, with some indirection going on in the magic cache due to the proxying taking place. In my tests the download speed (in term of overall timing) was pretty much identical though (tested from a few MiB to a few GiB). Could you tell me the durations in each case? Is it reproducible every time?

Thanks!

**Comment by crohr** [comment]
Also could you confirm the instance type that was used when using the magic cache? Was it the same (i.e. `c6in.8xlarge`) as for the runs-on/cache ?
